torch/csrc/Module.cpp

```diff
@@ -570,6 +573,118 @@ PyObject *THPModule_fromDLPack(PyObject *_unused, PyObject *data)
   return torch::createPyObject(atensor);
 }
 
+// In cases like data loader, if a worker process die due to bus error/segfault
+// or just hang, the main process, if implemented with multiprocessing.queue.,
+// will hang waiting for data. This is difficult to avoid on PyTorch side as it
+// can be caused by limited shm, other libraries users call in the workers. The
+// following methods is an effort to do our best provide some error message to
+// users when such unfortunate events happen.
+
+// Critical signal handlers should be registered on worker processes before
+// doing work.
+// Python handle is _set_worker_signal_handlers().
+#define SIGNAL_HANDLER(SIGNAL, HANDLER_NAME, ERROR_MSG)                       \
+static void HANDLER_NAME(int sig)                                             \
+{                                                                             \
+    write(fileno(stderr), ERROR_MSG, strlen(ERROR_MSG));                      \
```

Comment:
I don't think `fileno` is async-signal-safe:

http://man7.org/linux/man-pages/man7/signal-safety.7.html

Reply:
Ah... I checked most methods except fileno :P. Good catch!

Reply:
`strlen` isn't officially async-signal-safe either. If `ERROR_MSG` is going to be a literal, you can define a char array in this function and just use `sizeof` to get the length